File: src/parser.py
from src.lexer import Lexer
from src.tokens import Token, TokenType
from src.ast import Program, LetStatement, ConstStatement, Identifier, IntegerLiteral, FloatLiteral, StringLiteral, Expression, Statement

# Define operator precedence levels
LOWEST = 1
EQUALS = 2  # =
SUM = 3     # + or -
PRODUCT = 4 # * or /

class Parser:
    def __init__(self, lexer: Lexer):
        self.l = lexer
        self.errors: list[str] = []
        
        self.cur_token: Token = None
        self.peek_token: Token = None
        
        # Read two tokens so cur_token and peek_token are both set
        self.next_token()
        self.next_token()

    def next_token(self):
        """Advances both the current and peek tokens."""
        self.cur_token = self.peek_token
        self.peek_token = self.l.next_token()

    def cur_token_is(self, t: TokenType) -> bool:
        return self.cur_token.type == t

    def peek_token_is(self, t: TokenType) -> bool:
        return self.peek_token.type == t

    def expect_peek(self, t: TokenType) -> bool:
        """Enforces grammar structure. If the next token matches, advance. Otherwise, log an error."""
        if self.peek_token_is(t):
            self.next_token()
            return True
        else:
            self.peek_error(t)
            return False

    def peek_error(self, t: TokenType):
        msg = f"Line {self.peek_token.line}, Col {self.peek_token.column}: Expected next token to be {t.name}, got {self.peek_token.type.name} instead"
        self.errors.append(msg)

    def parse_program(self) -> Program:
        """Root parsing loop that structures the tokens into a program node."""
        program = Program()
        
        while not self.cur_token_is(TokenType.EOF):
            stmt = self.parse_statement()
            if stmt is not None:
                program.statements.append(stmt)
            self.next_token()
            
        return program

    def parse_statement(self) -> Statement:
        """Determines the path of statement parsing based on the current token."""
        if self.cur_token_is(TokenType.LET):
            return self.parse_let_statement()
        elif self.cur_token_is(TokenType.CONST):
            return self.parse_const_statement()
        else:
            # For Phase 1, we only focus on let/const declarations.
            # Unhandled standalone tokens are skipped or treated as invalid statements.
            return None

    def parse_let_statement(self) -> LetStatement:
        """Parses: let <identifier> [ : <type> ] = <expression>"""
        token = self.cur_token # The 'let' token
        
        if not self.expect_peek(TokenType.IDENTIFIER):
            return None
            
        name = Identifier(self.cur_token, self.cur_token.literal)
        explicit_type = None
        
        # Check if there is an optional explicit type declaration (e.g., ': Int')
        if self.peek_token_is(TokenType.COLON):
            self.next_token() # Current token is now ':'
            if not self.expect_peek(TokenType.IDENTIFIER):
                return None
            explicit_type = Identifier(self.cur_token, self.cur_token.literal)
            
        if not self.expect_peek(TokenType.ASSIGN):
            return None
            
        self.next_token() # Skip the '=' to get to the expression value
        
        value = self.parse_expression(LOWEST)
        
        return LetStatement(token=token, name=name, explicit_type=explicit_type, value=value)

    def parse_const_statement(self) -> ConstStatement:
        """Parses: const <identifier> [ : <type> ] = <expression>"""
        token = self.cur_token # The 'const' token
        
        if not self.expect_peek(TokenType.IDENTIFIER):
            return None
            
        name = Identifier(self.cur_token, self.cur_token.literal)
        explicit_type = None
        
        if self.peek_token_is(TokenType.COLON):
            self.next_token()
            if not self.expect_peek(TokenType.IDENTIFIER):
                return None
            explicit_type = Identifier(self.cur_token, self.cur_token.literal)
            
        if not self.expect_peek(TokenType.ASSIGN):
            return None
            
        self.next_token()
        
        value = self.parse_expression(LOWEST)
        
        return ConstStatement(token=token, name=name, explicit_type=explicit_type, value=value)

    def parse_expression(self, precedence: int) -> Expression:
        """Dispatches expression routing based on literal sub-types."""
        if self.cur_token_is(TokenType.IDENTIFIER):
            return Identifier(self.cur_token, self.cur_token.literal)
        elif self.cur_token_is(TokenType.INT_LIT):
            return IntegerLiteral(self.cur_token, int(self.cur_token.literal))
        elif self.cur_token_is(TokenType.FLOAT_LIT):
            return FloatLiteral(self.cur_token, float(self.cur_token.literal))
        elif self.cur_token_is(TokenType.STRING_LIT):
            return StringLiteral(self.cur_token, self.cur_token.literal)
        else:
            self.errors.append(f"Line {self.cur_token.line}: No expression parsing function found for {self.cur_token.type.name}")
            return None

File: src/parser.rs
use crate::lexer::Lexer;
use crate::tokens::{Token, TokenType};
use crate::ast::{Program, Statement, LetStatement, ConstStatement, Expression};

pub struct Parser {
    lexer: Lexer,
    cur_token: Token,
    peek_token: Token,
    pub errors: Vec<String>,
}

impl Parser {
    pub fn new(mut lexer: Lexer) -> Self {
        let cur_token = lexer.next_token();
        let peek_token = lexer.next_token();
        
        Parser {
            lexer,
            cur_token,
            peek_token,
            errors: Vec::new(),
        }
    }

    fn next_token(&mut self) {
        self.cur_token = self.peek_token.clone();
        self.peek_token = self.lexer.next_token();
    }

    fn cur_token_is(&self, t: TokenType) -> bool {
        self.cur_token.token_type == t
    }

    fn peek_token_is(&self, t: TokenType) -> bool {
        self.peek_token.token_type == t
    }

    fn expect_peek(&mut self, t: TokenType) -> bool {
        if self.peek_token_is(t) {
            self.next_token();
            true
        } else {
            self.peek_error(t);
            false
        }
    }

    fn peek_error(&mut self, t: TokenType) {
        let msg = format!(
            "Line {}, Col {}: Expected next token to be {:?}, got {:?} instead",
            self.peek_token.line, self.peek_token.column, t, self.peek_token.token_type
        );
        self.errors.push(msg);
    }

    pub fn parse_program(&mut self) -> Program {
        let mut program = Program { statements: Vec::new() };

        while !self.cur_token_is(TokenType::Eof) {
            if let Some(stmt) = self.parse_statement() {
                program.statements.push(stmt);
            }
            self.next_token();
        }

        program
    }

    fn parse_statement(&mut self) -> Option<Statement> {
        match self.cur_token.token_type {
            TokenType::Let => self.parse_let_statement().map(Statement::Let),
            TokenType::Const => self.parse_const_statement().map(Statement::Const),
            _ => None,
        }
    }

    fn parse_let_statement(&mut self) -> Option<LetStatement> {
        let token = self.cur_token.clone();

        if !self.expect_peek(TokenType::Identifier) {
            return None;
        }
        let name = self.cur_token.literal.clone();
        let mut explicit_type = None;

        if self.peek_token_is(TokenType::Colon) {
            self.next_token();
            if !self.expect_peek(TokenType::Identifier) {
                return None;
            }
            explicit_type = Some(self.cur_token.literal.clone());
        }

        if !self.expect_peek(TokenType::Assign) {
            return None;
        }
        self.next_token();

        let value = self.parse_expression()?;

        Some(LetStatement {
            token,
            name,
            explicit_type,
            value,
        })
    }

    fn parse_const_statement(&mut self) -> Option<ConstStatement> {
        let token = self.cur_token.clone();

        if !self.expect_peek(TokenType::Identifier) {
            return None;
        }
        let name = self.cur_token.literal.clone();
        let mut explicit_type = None;

        if self.peek_token_is(TokenType::Colon) {
            self.next_token();
            if !self.expect_peek(TokenType::Identifier) {
                return None;
            }
            explicit_type = Some(self.cur_token.literal.clone());
        }

        if !self.expect_peek(TokenType::Assign) {
            return None;
        }
        self.next_token();

        let value = self.parse_expression()?;

        Some(ConstStatement {
            token,
            name,
            explicit_type,
            value,
        })
    }

    fn parse_expression(&mut self) -> Option<Expression> {
        match self.cur_token.token_type {
            TokenType::Identifier => Some(Expression::Identifier(self.cur_token.literal.clone())),
            TokenType::IntLit => self.cur_token.literal.parse::<i64>().ok().map(Expression::IntegerLiteral),
            TokenType::FloatLit => self.cur_token.literal.parse::<f64>().ok().map(Expression::FloatLiteral),
            TokenType::StringLit => Some(Expression::StringLiteral(self.cur_token.literal.clone())),
            _ => {
                self.errors.push(format!("Line {}: No expression parser match found", self.cur_token.line));
                None
            }
        }
    }
}
